In [2]:
import pandas as pd
df=pd.read_csv("D:\\New folder\\food_coded.csv")
df

GPA  Gender  breakfast  calories_chicken  calories_day  calories_scone  \
0      2.4       2          1               430           NaN           315.0   
1    3.654       1          1               610           3.0           420.0   
2      3.3       1          1               720           4.0           420.0   
3      3.2       1          1               430           3.0           420.0   
4      3.5       1          1               720           2.0           420.0   
..     ...     ...        ...               ...           ...             ...   
120    3.5       1          1               610           4.0           420.0   
121      3       1          1               265           2.0           315.0   
122  3.882       1          1               720           NaN           420.0   
123      3       2          1               720           4.0           420.0   
124    3.9       1          1               430           NaN           315.0   

     coffee                             comfort_food  \
0         1                                     none   
1         2              chocolate, chips, ice cream   
2         2          frozen yogurt, pizza, fast food   
3         2         Pizza, Mac and cheese, ice cream   
4         2             Ice cream, chocolate, chips    
..      ...                                      ...   
120       2  wine. mac and cheese, pizza, ice cream    
121       2               Pizza / Wings / Cheesecake   
122       1               rice, potato, seaweed soup   
123       1             Mac n Cheese, Lasagna, Pizza   
124       2             Chocolates, pizza, and Ritz.   

                              comfort_food_reasons  \
0                            we dont have comfort    
1                             Stress, bored, anger   
2                                  stress, sadness   
3                                          Boredom   
4                       Stress, boredom, cravings    
..                                             ...   
120                           boredom and sadness    
121                Loneliness / Homesick / Sadness   
122                                        sadness   
123  happiness, they are some of my favorite foods   
124               hormones, Premenstrual syndrome.   

     comfort_food_reasons_coded  ...  soup  sports  thai_food  \
0                           9.0  ...   1.0     1.0          1   
1                           1.0  ...   1.0     1.0          2   
2                           1.0  ...   1.0     2.0          5   
3                           2.0  ...   1.0     2.0          5   
4                           1.0  ...   1.0     1.0          4   
..                          ...  ...   ...     ...        ...   
120                         NaN  ...   1.0     1.0          5   
121                         NaN  ...   1.0     NaN          4   
122                         NaN  ...   1.0     2.0          5   
123                         NaN  ...   2.0     2.0          1   
124                         NaN  ...   1.0     2.0          2   

    tortilla_calories  turkey_calories  type_sports veggies_day  vitamins  \
0              1165.0              345   car racing           5         1   
1               725.0              690  Basketball            4         2   
2              1165.0              500         none           5         1   
3               725.0              690          NaN           3         1   
4               940.0              500     Softball           4         2   
..                ...              ...          ...         ...       ...   
120             940.0              500     Softball           5         1   
121             940.0              500  basketball            5         2   
122             580.0              690         none           4         2   
123             940.0              500          NaN           3         1   
124             725.0              345          NaN           4         2   

     w

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125 entries, 0 to 124
Data columns (total 61 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   GPA                           123 non-null    object 
 1   Gender                        125 non-null    int64  
 2   breakfast                     125 non-null    int64  
 3   calories_chicken              125 non-null    int64  
 4   calories_day                  106 non-null    float64
 5   calories_scone                124 non-null    float64
 6   coffee                        125 non-null    int64  
 7   comfort_food                  124 non-null    object 
 8   comfort_food_reasons          123 non-null    object 
 9   comfort_food_reasons_coded    106 non-null    float64
 10  cook                          122 non-null    float64
 11  comfort_food_reasons_coded.1  125 non-null    int64  
 12  cuisine                       108 non-null    float64
 13  diet_

In [4]:
df.isnull().sum()

GPA                  2
Gender               0
breakfast            0
calories_chicken     0
calories_day        19
                    ..
type_sports         26
veggies_day          0
vitamins             0
waffle_calories      0
weight               2
Length: 61, dtype: int64

In [5]:
df.duplicated().sum()

0

In [6]:
df.describe

<bound method NDFrame.describe of        GPA  Gender  breakfast  calories_chicken  calories_day  calories_scone  \
0      2.4       2          1               430           NaN           315.0   
1    3.654       1          1               610           3.0           420.0   
2      3.3       1          1               720           4.0           420.0   
3      3.2       1          1               430           3.0           420.0   
4      3.5       1          1               720           2.0           420.0   
..     ...     ...        ...               ...           ...             ...   
120    3.5       1          1               610           4.0           420.0   
121      3       1          1               265           2.0           315.0   
122  3.882       1          1               720           NaN           420.0   
123      3       2          1               720           4.0           420.0   
124    3.9       1          1               430           NaN           315

Handling missing values

->methods for handling missing values

1. **Complete Case Analysis (CCA)**:
   - In complete case analysis, also known as listwise deletion, you simply remove any observations from your dataset that contain missing values in any of the columns.
   - This method is straightforward and easy to implement.
   - However, it can lead to loss of valuable information, especially if the missing values are not missing completely at random (MCAR).

2. **Pairwise Deletion**:
   - With pairwise deletion, also known as available case analysis, you retain observations with missing values for specific analyses but only use the available data for each calculation.
   - This method allows you to maximize the use of available data without discarding entire observations.
   - However, it can lead to biased estimates if missingness is related to the variables being analyzed.

3. **Threshold-based Removal**:
   - In threshold-based removal, you set a threshold for the proportion of missing values allowed in each observation or column. Observations or columns exceeding this threshold are removed from the dataset.
   - This method allows you to control the extent of data loss by adjusting the threshold.
   - However, choosing an appropriate threshold can be subjective and may require domain knowledge or sensitivity analysis.

4. **Column-wise Removal**:
   - If certain columns in your dataset have a high proportion of missing values and are not critical for your analysis, you may choose to remove those columns entirely.
   - This method can help reduce the complexity of your dataset and focus your analysis on the most informative variables.
   - However, it may result in the loss of potentially relevant information, so it's essential to carefully evaluate the importance of each column.

5. **Row-wise Removal**:
   - Similarly, if certain observations have missing values across multiple columns and cannot be imputed accurately, you may choose to remove those observations entirely.
   - This method can help ensure the quality of your dataset by eliminating potentially unreliable or biased observations.
   - However, it may reduce the sample size and statistical power of your analysis, particularly if missingness is non-random.

6. **Imputation before removal**:
   - Before removing missing values, you may consider imputing missing values for certain columns to retain as much information as possible.
   - After imputation, you can then apply one of the aforementioned methods to remove remaining missing values or observations.
   - This approach allows you to balance the trade-off between data completeness and bias reduction.

Here dataset size  is small (125 rows), and considering the potential loss of valuable information,we can use the following approaches:

1. **Imputation before removal**: Since you have a small dataset, imputing missing values before removing them might be beneficial. You can use appropriate imputation methods (e.g., mean, median, mode imputation for numerical or categorical variables) to fill in missing values, thereby retaining as much information as possible.

2. **Threshold-based Removal**: Considering the small number of observations, you might set a relatively lenient threshold for removing observations or columns with missing values. This approach allows you to control the extent of data loss while ensuring that the remaining dataset is relatively complete.

Combining these approaches can help mitigate the loss of information while still addressing the issue of missing values. Here's how you might proceed:

- **Impute missing values**: Use imputation methods to fill in missing values in your dataset.
- **Threshold-based removal**: After imputation, apply a threshold-based approach to remove any remaining observations or columns with a high proportion of missing values.




In [7]:
numeric_columns = []
for column in df.columns:
    # Check if the column contains numeric values
    if pd.api.types.is_numeric_dtype(df[column]):
        # If yes, append the column name to the list
        numeric_columns.append(column)

# Print the list of numeric column names
print("Numeric columns:", numeric_columns)

Numeric columns: ['Gender', 'breakfast', 'calories_chicken', 'calories_day', 'calories_scone', 'coffee', 'comfort_food_reasons_coded', 'cook', 'comfort_food_reasons_coded.1', 'cuisine', 'diet_current_coded', 'drink', 'eating_changes_coded', 'eating_changes_coded1', 'eating_out', 'employment', 'ethnic_food', 'exercise', 'father_education', 'fav_cuisine_coded', 'fav_food', 'fries', 'fruit_day', 'grade_level', 'greek_food', 'healthy_feeling', 'ideal_diet_coded', 'income', 'indian_food', 'italian_food', 'life_rewarding', 'marital_status', 'mother_education', 'nutritional_check', 'on_off_campus', 'parents_cook', 'pay_meal_out', 'persian_food', 'self_perception_weight', 'soup', 'sports', 'thai_food', 'tortilla_calories', 'turkey_calories', 'veggies_day', 'vitamins', 'waffle_calories']


In [8]:
numeric_columns = df.select_dtypes(include=['number']).columns

# Apply mean imputation to numeric columns
for column in numeric_columns:
    # Calculate the mean of the column
    column_mean = df[column].mean()
    # Fill missing values in the column with the mean
    df[column].fillna(column_mean, inplace=True)

# Print the DataFrame after mean imputation
print(df)

       GPA  Gender  breakfast  calories_chicken  calories_day  calories_scone  \
0      2.4       2          1               430      3.028302           315.0   
1    3.654       1          1               610      3.000000           420.0   
2      3.3       1          1               720      4.000000           420.0   
3      3.2       1          1               430      3.000000           420.0   
4      3.5       1          1               720      2.000000           420.0   
..     ...     ...        ...               ...           ...             ...   
120    3.5       1          1               610      4.000000           420.0   
121      3       1          1               265      2.000000           315.0   
122  3.882       1          1               720      3.028302           420.0   
123      3       2          1               720      4.000000           420.0   
124    3.9       1          1               430      3.028302           315.0   

     coffee                

In [9]:
df.isnull().sum()

GPA                  2
Gender               0
breakfast            0
calories_chicken     0
calories_day         0
                    ..
type_sports         26
veggies_day          0
vitamins             0
waffle_calories      0
weight               2
Length: 61, dtype: int64